### DISCUSSION TITLE
Specifying array types when generating migrations

### DISCUSSION THREAD

**Discussion by mateusdeap** [loco-rs/loco#301] [Q&A] [answered]
So, according to SeaORM's [docs](https://www.sea-ql.org/SeaORM/docs/generate-entity/entity-structure/#column-type), in order to specify a column is an array (assuming you're using Postgres), you just declare the struct field as being a `Vec` of known rust types, like so:

```rust
#[derive(Clone, Debug, PartialEq, DeriveEntityModel)]
#[sea_orm(table_name = "collection")]
pub struct Model {
    #[sea_orm(primary_key)]
    pub id: i32,
    pub integers: Vec<i32>,
    pub integers_opt: Option<Vec<i32>>,
    pub floats: Vec<f32>,
    pub doubles: Vec<f64>,
    pub strings: Vec<String>,
}
```

I tried searching the docs and couldn't find the list of types we can specify when generating a migration. Am I correct in assuming that specifying columns as arrays of ints, etc. still hasn't been implemented in the CLI?

**Comment by futuretim** [comment, marked as answer]
It sounds like you are having a similar issue I was having with floats. I couldn't find documentation on how to use them as the CLI registry/schema does not appear to offer that mapping when generating.

The same appears to be the case for Vec. However, if you just use underlying SeaORM logic you'll have no problem. So once you have a migration ready you can modify it. Let's say in the migration you generated you have a single integer value like so:

```
async fn up(&self, manager: &SchemaManager) -> Result<(), DbErr> {
        manager
            .create_table(
                table_auto(SomeModel::Table)
                   .col(integer(SomeModel::MyValue).borrow_mut())
                  .to_owned(),
)
.await
}
```

You can change that like so:

```
async fn up(&self, manager: &SchemaManager) -> Result<(), DbErr> {
        manager
            .create_table(
                table_auto(SomeModel::Table)
                   .col(ColumnDef::new(SomeModel::MyValue).array(ColumnType::Integer).not_null())
                  .to_owned(),
)
.await
}
```

Now you need to run `cargo loco db migrate` and then `cargo loco db entities` and you should get the result you want. Let me know if that works out for you.

**Reply by mateusdeap** [reply]
So I went through SeaORM's documentation and came to the same conclusion.

I'm going to give your solution a try. If it works, I'm going to try and add a function for this.

**Reply by mateusdeap** [reply]
@futuretim It seems to have worked. the commands ran fine, but I do have some questions.

1. The array field in the entity struct generated had this: `pub winning_numbers: Option<i32>`. Shouldn't it be a `Vec<i32>`?
2. Why do I change the last function call from `borrow_mut()` to `not_null()`? I understand borrow_mut probably isn't implemented for the ColumnDef struct it returns, but shouldn't I tack it after the call to `not_null()`?

**Comment by futuretim** [comment]
Yes. Something did not work correctly if that was the outcome. Presumably you got your database in a state to where the migration ran again after the modification, right?

I believe the difference is that when you apply the modifiers with SeaORM you get back a mutable reference. I *think* loco's schema modifiers could return mutable references and it would be equivalent, they just don't.

**Comment by mateusdeap** [comment]
Just to update: I just dropped my database and recreated it. It all worked.
